heutagogy/auth.py

```diff
@@ -1,34 +1,99 @@
+
+First one is JWT tokens and is main authentication method used by all
+API methods.
+
+Second method is session-based and is used by local user-management
+interface. (The pages for user registration, log-in, change username
+and password.)
+
+This module provides integration of Flask-JWT and Flask-User.
+"""
 from heutagogy import app
-from heutagogy.persistence import User
-from flask_jwt import JWT, jwt_required, current_identity
+from heutagogy.persistence import db
+from flask_jwt import JWT, JWTError, jwt_required, current_identity
 import flask_login
+import flask_user
+
+
+class User(db.Model, flask_user.UserMixin):
+    id = db.Column(db.Integer, primary_key=True)
+    username = db.Column(db.String(255), nullable=False, unique=True)
+    password = db.Column(db.String, nullable=False)
+
+    email = db.Column(db.String(255), nullable=False, unique=True)
+    confirmed_at = db.Column(db.DateTime())
+
+    active = db.Column('is_active', db.Boolean(), nullable=False,
+                       server_default='0')
+    first_name = db.Column(db.String(255), nullable=False, server_default='')
+    last_name = db.Column(db.String(255), nullable=False, server_default='')
 
+    def set_password(self, password):
+        self.password = user_manager.hash_password(password)
 
-def get_user(username):
-    return User.query.filter_by(username=username).first()
+
+def get_user(user_id):
+    return user_manager.get_user_by_id(user_id)
 
 
 def authenticate(username, password):
-    user = get_user(username)
-    if user and user.check_password(password):
+    user = user_manager.find_user_by_username(username)
+    if user and user_manager.verify_password(password, user):
         return user
 
 
 def identity(payload):
     user = get_user(payload['identity'])
-    if user:
-        # this field is needed for Flask-Login
-        user.is_authenticated = True
     return user
 
 
 jwt = JWT(app, authenticate, identity)
-
-
-login_manager = flask_login.LoginManager(app)
+login_manager = flask_login.LoginManager()
+db_adapter = flask_user.SQLAlchemyAdapter(db, User)
+user_manager = flask_user.UserManager(db_adapter, app,
+                                      login_manager=login_manager)
 
 
 @login_manager.request_loader
-@jwt_required()
 def load_user_from_request(request):
-    return current_identity
+    """This function is called by login manager to get user info from
+    the request (JWT authorization) when there is no info in the
+    session.
+
+    The trick here is that we don't want it to error when there is no
+    JWT token, as otherwise the user can never access registration
+    page.
+    """
+    @jwt_required()
+    def get_user():
+        return current_identity
+
+    try:
+        return get_user()
+    except JWTError as e:
+        # Don't error when no JWT token is provided
+        if e.error == 'Authorization Required':
+            return login_manager.anonymous_user()
+        else:
+            raise
```

Comment:
What do you think about checking a request URL and returning `login_manager.anonymous_user()` only if there is `/user/register` in it, otherwise `get_user()`?